### ISSUE TITLE
Unstable nn.MultiLabelSoftMarginCriterion

### ISSUE THREAD

**Issue by arunmallya** [torch/nn#949]
Shouldn't the Sigmoid and Binary Cross Entropy functions be combined inside nn.MultiLabelSoftMarginCriterion for numerical stability? 

The current implementation calls Sigmoid followed by BCE on the inputs, and the reverse for gradient computation. However, when activations are large, the results go all wrong.

```
x = torch.Tensor{-50}
t = torch.Tensor{1}

mls = nn.MultiLabelSoftMarginCriterion()
mls_loss = mls:forward(x, t)
> 27.63

mls_grad = mls:backward(x, t)
>
1e-10 *
-1.9287
[torch.DoubleTensor of size 1]

```

However, the numerically stable loss/gradient equation as described in https://www.ics.uci.edu/~pjsadows/notes.pdf (Eq 1-10) calculates the gradient 
to be **-1** and not **1e-10 *-1.9287**.
The dirty denominator which causes problems gets cancelled out in Eqs. 6 and 7. The stable equations are also utilized in caffe: https://github.com/BVLC/caffe/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L42

**Comment by soumith** [comment]
hi @arunmallya . This looks like it is indeed unstable.
nn.MultiLabelSoftMarginCriterion can be short-cut to be much more stable. 
Are you willing to send a PR for this? if not i will get to it in due time.

**Comment by arunmallya** [comment]
Yeah, will do.

**Comment by arunmallya** [comment]
Submitted pull request: https://github.com/torch/nn/pull/953
